### PR TITLE
sending correct external resources for k8s-array plugin

### PR DESCRIPTION
**Pull request by hamersaw**
# TL;DR
Cache status of map tasks subtasks is not reported correctly. This PR fixes the issue, where all external resources may be correctly annotated with the correct cache status during execution.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
_NA_

## Follow-up issue
_NA_

### CODE REVIEW

File: go/tasks/plugins/array/catalog.go [hunk 1]
```diff
@@ -257,18 +257,21 @@ func WriteToDiscovery(ctx context.Context, tCtx core.TaskExecutionContext, state
...
 	// Do not cache failed tasks. Retrieve the final phase from array status and unset the non-successful ones.
+
 	tasksToCache := state.GetIndexesToCache().DeepCopy()
 	for idx, phaseIdx := range state.ArrayStatus.Detailed.GetItems() {
 		phase := core.Phases[phaseIdx]
 		if !phase.IsSuccess() {
-			tasksToCache.Clear(uint(idx))
+			// tasksToCache is built on the originalArraySize and ArrayStatus.Detailed is the executionArraySize
+			originalIdx := arrayCore.CalculateOriginalIndex(idx, state.GetIndexesToCache())
+			tasksToCache.Clear(uint(originalIdx))
```
Comment (EngHabu) [inline go/tasks/plugins/array/catalog.go:274]: Ouch

File: go/tasks/plugins/array/core/state.go [hunk 2]
```diff
@@ -267,8 +263,9 @@ func SummaryToPhase(ctx context.Context, minSuccesses int64, summary arraystatus
 		return PhaseWriteToDiscoveryThenFail
 	}
 
-	// No chance to reach the required success numbers.
-	if totalRunning+totalSuccesses+totalWaitingForResources+totalRetryableFailures < minSuccesses {
+	// No chance to reach the required success numbers. If any subtasks are running we let them
+	// complete, so results may be written to cache if enabled, before failing.
+	if totalSuccesses+totalWaitingForResources+totalRetryableFailures < minSuccesses && totalRunning == 0 {
```
Comment (EngHabu) [inline go/tasks/plugins/array/core/state.go:268]: Should we just add total running to the +s

Comment (hamersaw) [reply go/tasks/plugins/array/core/state.go]: OK, so I felt strongly enough to change it, but not enough to defend the change. I reverted this to the previous approach. Basically, I think transitioning to permanent failure should operate differently whether the task is cached or not:
Cached: execute as many of the subtasks as possible to allow the cached outputs to be written.
Not cached: fail as soon as possible so you don't spend any time doing unnecessary work (current approach).

**Comment by cosmicBboy** [comment]
@EngHabu friendly ping: does this PR look good to go? Currently blocking https://github.com/flyteorg/flyte/issues/3315

**Pull request merged**